### ISSUE TITLE
Generating Homography and Inverse Homography

### ISSUE THREAD

**Issue by MuhamedEssam** [mguti97/No-Bells-Just-Whistles#4]
For Analysis purpose, I would like to generate homography and inverse homography. I already tried to allocate points from  3D Field points to Projected 2D of live cam, to generate homography and inverse homography. but it didn't work well. do you have any idea? to help so i can draw parallel lines to goal lines or to touch lines  in the 2d live cam ?

**Comment by mguti97** [comment]
I can't help you without extra information. Feel free to bring your code and outputs and I'd be glad to help.

**Comment by mguti97** [comment]
I can't help you without extra information. Feel free to bring your code and outputs and I'd be glad to help.

**Comment by mguti97** [comment]
I can't help you without extra information. Feel free to bring your code and outputs and I'd be glad to help.

**Comment by MuhamedEssam** [comment]
`import cv2
import sys
import yaml
import torch
import argparse
import numpy as np
import matplotlib.pyplot as plt
import torchvision.transforms as T
import torchvision.transforms.functional as f

from tqdm import tqdm
from PIL import Image
from matplotlib.patches import Polygon


from model.cls_hrnet import get_cls_net
from model.cls_hrnet_l import get_cls_net as get_cls_net_l

from utils.utils_calib import FramebyFrameCalib, pan_tilt_roll_to_orientation
from utils.utils_heatmap import get_keypoints_from_heatmap_batch_maxpool, get_keypoints_from_heatmap_batch_maxpool_l, \
    complete_keypoints, coords_to_dict


lines_coords = [[[0., 54.16, 0.], [16.5, 54.16, 0.]],
                [[16.5, 13.84, 0.], [16.5, 54.16, 0.]], # mark line
                [[16.5, 13.84, 0.], [0., 13.84, 0.]],
                [[88.5, 54.16, 0.], [105., 54.16, 0.]],
                [[88.5, 13.84, 0.], [88.5, 54.16, 0.]], # mark line 2
                [[88.5, 13.84, 0.], [105., 13.84, 0.]],
                [[0., 37.66, -2.44], [0., 30.34, -2.44]],
                [[0., 37.66, 0.], [0., 37.66, -2.44]],
                [[0., 30.34, 0.], [0., 30.34, -2.44]],
                [[105., 37.66, -2.44], [105., 30.34, -2.44]],
                [[105., 30.34, 0.], [105., 30.34, -2.44]],
                [[105., 37.66, 0.], [105., 37.66, -2.44]],
                [[52.5, 0., 0.], [52.5, 68, 0.]],
                [[0., 68., 0.], [105., 68., 0.]],
                [[0., 0., 0.], [0., 68., 0.]],
                [[105., 0., 0.], [105., 68., 0.]],
                [[0., 0., 0.], [105., 0., 0.]],
                [[0., 43.16, 0.], [5.5, 43.16, 0.]],
                [[5.5, 43.16, 0.], [5.5, 24.84, 0.]],
                [[5.5, 24.84, 0.], [0., 24.84, 0.]],
                [[99.5, 43.16, 0.], [105., 43.16, 0.]],
                [[99.5, 43.16, 0.], [99.5, 24.84, 0.]],
                [[99.5, 24.84, 0.], [105., 24.84, 0.]]]


def projection_from_cam_params(final_params_dict):
    cam_params = final_params_dict["cam_params"]
    x_focal_length = cam_params['x_focal_length']
    y_focal_length = cam_params['y_focal_length']
    principal_point = np.array(cam_params['principal_point'])
    position_meters = np.array(cam_params['position_meters'])
    rotation = np.array(cam_params['rotation_matrix'])

    It = np.eye(4)[:-1]
    It[:, -1] = -position_meters
    Q = np.array([[x_focal_length, 0, principal_point[0]],
                  [0, y_focal_length, principal_point[1]],
                  [0, 0, 1]])
    P = Q @ (rotation @ It)

    return P

def generate_random_points(num_points=10):
    """ Generate 10 random points on the field """
    points = []
    for _ in range(num_points):
        x = np.random.uniform(0, 105)  # Width of the field
        y = np.random.uniform(0, 68)   # Height of the field
        points.append([x, y, 0., 1.])  # 3D points with z=0, homogeneous coordinates
    return np.array(points)

def get_homography_and_inverse(points, projected_points):
  
    """ Verify homography matrix by transforming points and back-transforming them """
    # Compute the homography matrix
    h_matrix, _ = cv2.findHomography(points[:, :2], projected_points, method=0)

    # Apply the homography transformation
    transformed_points = cv2.perspectiveTransform(points[:, :2].reshape(-1, 1, 2), h_matrix)

    # Compute the inverse homography matrix
    h_inv_matrix = np.linalg.inv(h_matrix)

    # Apply the inverse homography transformation
    inverse_transformed_points = cv2.perspectiveTransform(transformed_points, h_inv_matrix)

    # Compare original points and inverse transformed points to verify
    for i in range(len(points)):
        print(f"Original: {points[i][:2]}, After transformation and inverse: {inverse_transformed_points[i][0]}")
        assert np.allclose(points[i][:2], inverse_transformed_points[i][0]), "Point mismatch!"

    print("Homography verification successful: All points match after round-trip transformation.")
    return h_matrix , h_inv_matrix

def transform_point(point, inv_homography_matrix):
    # Transform the (500,500) point using the inverse homography matrix
    p = np.array([point[0], point[1], 1])
    transformed_p = inv_homography_matrix @ p
    transformed_p /= transformed_p[-1]  # Normalize
    return transformed_p

def project_points(P, points):
    """ Project 3D points using matrix P """
    projected_points = []
    for point in points:
        projected = P @ point
        projected /= projected[-1]  # Normalize to get the 2D point
        projected_points.append([projected[0], projected[1]])
    return np.array(projected_points)

def project_and_verify(P): 
    # Generate 10 random field points
    random_points = generate_random_points(10)
    # Project these points using P
    projected_points = project_points(P, random_points)
    # Verify homography
    h_matrix , h_inv_matrix= get_homography_and_inverse(random_points, projected_points)
    return h_matrix , h_inv_matrix

def inference(cam, frame, model, model_l, kp_threshold, line_threshold):
    frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    frame = Image.fromarray(frame)
    frame = f.to_tensor(frame).float().unsqueeze(0)
    _, _, h_original, w_original = frame.size()
    frame = frame if frame.size()[-1] == 960 else transform2(frame)
    frame = frame.to(device)
    b, c, h, w = frame.size()
    # print (type(frame))  # <class 'torch.Tensor'>
    # print(frame)
    # print(model)
    with torch.no_grad():
        heatmaps = model(frame)
        heatmaps_l = model_l(frame)
    kp_coords = get_keypoints_from_heatmap_batch_maxpool(heatmaps[:,:-1,:,:])
    line_coords = get_keypoints_from_heatmap_batch_maxpool_l(heatmaps_l[:,:-1,:,:])
    kp_dict = coords_to_dict(kp_coords, threshold=kp_threshold)
    lines_dict = coords_to_dict(line_coords, threshold=line_threshold)
    final_dict = complete_keypoints(kp_dict, lines_dict, w=w, h=h, normalize=True)
    cam.update(final_dict[0])
    final_params_dict = cam.heuristic_voting()
    return final_params_dict


def project(frame, P ,offside1=None,offside2=None):
    for line in lines_coords:
        w1 = line[0]
        w2 = line[1]
        i1 = P @ np.array([w1[0]-105/2, w1[1]-68/2, w1[2], 1])
        i2 = P @ np.array([w2[0]-105/2, w2[1]-68/2, w2[2], 1])
        i1 /= i1[-1]
        i2 /= i2[-1]
        frame = cv2.line(frame, (int(i1[0]), int(i1[1])), (int(i2[0]), int(i2[1])), (255, 0, 0), 3)
    r = 9.15
    pts1, pts2, pts3 = [], [], []
    base_pos = np.array([11-105/2, 68/2-68/2, 0., 0.])
    for ang in np.linspace(37, 143, 50):
        ang = np.deg2rad(ang)
        pos = base_pos + np.array([r*np.sin(ang), r*np.cos(ang), 0., 1.])
        ipos = P @ pos
        ipos /= ipos[-1]
        pts1.append([ipos[0], ipos[1]])
    base_pos = np.array([94-105/2, 68/2-68/2, 0., 0.])
    for ang in np.linspace(217, 323, 200):
        ang = np.deg2rad(ang)
        pos = base_pos + np.array([r*np.sin(ang), r*np.cos(ang), 0., 1.])
        ipos = P @ pos
        ipos /= ipos[-1]
        pts2.append([ipos[0], ipos[1]])
    base_pos = np.array([0, 0, 0., 0.])
    for ang in np.linspace(0, 360, 500):
        ang = np.deg2rad(ang)
        pos = base_pos + np.array([r*np.sin(ang), r*np.cos(ang), 0., 1.])
        ipos = P @ pos
        ipos /= ipos[-1]
        pts3.append([ipos[0], ipos[1]])
    XEllipse1 = np.array(pts1, np.int32)
    XEllipse2 = np.array(pts2, np.int32)
    XEllipse3 = np.array(pts3, np.int32)
    frame = cv2.polylines(frame, [XEllipse1], False, (255, 0, 0), 3)
    frame = cv2.polylines(frame, [XEllipse2], False, (255, 0, 0), 3)
    frame = cv2.polylines(frame, [XEllipse3], False, (255, 0, 0), 3)
    if offside1 is not None:
        print (offside1[0][0][0])
        print (offside1[0][0][1])
        w1 = np.array([offside1[0][0][0] - 105/2, 0 - 68/2, 0, 1])  # Offside line start point 
        w2 = np.array([offside1[0][0][0] - 105/2, 68 - 68/2, 0, 1])  # Offside line end point 
        w3 = np.array([offside1[0][0][0] - 105/2, offside1[0][0][1] - 68/2, 0, 1])  # Offside point 
        i1 = P @ w1
        i2 = P @ w2
        i3 = P @ w3
        i1 /= i1[-1]
        i2 /= i2[-1]
        i3 /= i3[-1]
        # Draw the offside line in yellow
        print('offside line:', i1)
        print('offside line 2:',i2)
        print('offside:',i3)
        frame = cv2.line(frame, (int(i1[0]), int(i1[1])), (int(i2[0]), int(i2[1])), (0, 255, 255), 3)
        frame = cv2.circle(frame, (int(i3[0]), int(i3[1])), 5, (0, 0, 255), -1)
    if offside2 is not None:
        w1 = np.array([offside2[0] - 105/2, 0 - 68/2, 0, 1])  # Offside line point 1
        w2 = np.array([offside2[0] - 105/2, 68 - 68/2, 0, 1])  # Offside line point 2
        w3 = np.array([offside2[0] - 105/2, offside2[1] - 68/2, 0, 1])  # Offside point 
        i1 = P @ w1
        i2 = P @ w2
        i3 = P @ w3
        i1 /= i1[-1]
        i2 /= i2[-1]
        i3 /= i3[-1]
        # Draw the offside line in yellow
        print('offside line:', i1)
        print('offside line 2:',i2)
        print('offside:',i3)
        frame = cv2.line(frame, (int(i1[0]), int(i1[1])), (int(i2[0]), int(i2[1])), (0, 255, 255), 3)
        frame = cv2.circle(frame, (int(i3[0]), int(i3[1])), 5, (0, 0, 255), -1)
    return frame


def process_input(input_path, input_type, model_kp, model_line, kp_threshold, line_threshold, save_path, display):
    cap = cv2.VideoCapture(input_path)
    frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    fps = int(cap.get(cv2.CAP_PROP_FPS))
    cam = FramebyFrameCalib(iwidth=frame_width, iheight=frame_height, denormalize=True)
    if input_type == 'image':
        frame = cv2.imread(input_path)
        if frame is None:
            print(f"Error: Unable to read the image {input_path}")
            return
        final_params_dict = inference(cam, frame, model, model_l, kp_threshold, line_threshold)
        if final_params_dict is not None:
            P = projection_from_cam_params(final_params_dict)
            print(' verifing')
            h_matrix , h_inv_matrix = project_and_verify(P)
            transformed_point = cv2.perspectiveTransform(np.array([[[348, ]]], dtype=np.float32), h_inv_matrix)
            print('transformed_point',transformed_point)
            print(transformed_point.shape)
            original= cv2.perspectiveTransform(transformed_point, h_matrix)
            print('original',original)
            print(original.shape)
            frame = cv2.circle(frame, (int(original[0][0][0]), int(original[0][0][1])), 5, (0, 0, 255), -1)
            x_value = transformed_point[0][0][0]  # Extract x-coordinate
            offside_pt_1,offside_pt_2 = (x_value -105/2, 0 -68/2 ) , (x_value -105/2 , 68 -68/2)
            print(offside_pt_1)
            print(offside_pt_2)
            transformed_pt_1 = cv2.perspectiveTransform(np.array([[offside_pt_1]], dtype=np.float32), h_matrix)
            transformed_pt_2 = cv2.perspectiveTransform(np.array([[offside_pt_2]], dtype=np.float32), h_matrix)
            print('transformed...')
            print(transformed_pt_1)
            print(transformed_pt_2)
            # Project and draw original soccer lines
            projected_frame = frame
            # # Project and draw original soccer lines
            projected_frame = project(frame, P , offside1=transformed_point)
        else:
            projected_frame = frame
        if save_path != "":
            cv2.imwrite(save_path, projected_frame)
        else:
            plt.imshow(cv2.cvtColor(projected_frame, cv2.COLOR_BGR2RGB))
            plt.axis('off')
            plt.show()
    return h_matrix, h_inv_matrix

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Process video or image and plot lines on each frame.")
    parser.add_argument("--weights_kp", type=str, help="Path to the model for keypoint inference.")
    parser.add_argument("--weights_line", type=str, help="Path to the model for line projection.")
    parser.add_argument("--kp_threshold", type=float, default=0.1486, help="Threshold for keypoint detection.")
    parser.add_argument("--line_threshold", type=float, default=0.3880, help="Threshold for line detection.")
    parser.add_argument("--device", type=str, default="cuda:0", help="CPU or CUDA device index")
    parser.add_argument("--input_path", type=str, required=True, help="Path to the input video or image file.")
    parser.add_argument("--input_type", type=str, choices=['video', 'image'], required=True, help="Type of input: 'video' or 'image'.")
    parser.add_argument("--save_path", type=str, default="", help="Path to save the processed video.")
    parser.add_argument("--display", action="store_true", help="Enable real-time display.")
    args = parser.parse_args()
`

**Comment by MuhamedEssam** [comment]
the core problem of this code lands here  in this function 
`
def get_homography_and_inverse(points, projected_points):      
      """ Verify homography matrix by transforming points and back-transforming them """
      # Compute the homography matrix
      h_matrix, _ = cv2.findHomography(points[:, :2], projected_points, method=0)
      
      # Apply the homography transformation
      transformed_points = cv2.perspectiveTransform(points[:, :2].reshape(-1, 1, 2), h_matrix)
      
      # Compute the inverse homography matrix
      h_inv_matrix = np.linalg.inv(h_matrix)
      
      # Apply the inverse homography transformation
      inverse_transformed_points = cv2.perspectiveTransform(transformed_points, h_inv_matrix)
      
      # Compare original points and inverse transformed points to verify
      for i in range(len(points)):
          print(f"Original: {points[i][:2]}, After transformation and inverse: {inverse_transformed_points[i][0]}")
          assert np.allclose(points[i][:2], inverse_transformed_points[i][0]), "Point mismatch!"
      
      print("Homography verification successful: All points match after round-trip transformation.")
      return h_matrix , h_inv_matrix
`
